File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    for i in range(1, len(triangle)):
        triangle[i][0] += triangle[i-1][0]
        triangle[i][-1] += triangle[i-1][-1]
        for j in range(1, i):
            if triangle[i-1][j-1] < triangle[i-1][j]: 
                triangle[i][j] += triangle[i-1][j]
            else:
                triangle[i][j] += triangle[i-1][j-1]
    return max(triangle[-1])

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;
        int N = triangle.length;
        int M = triangle[N-1].length;
        int[][] dp = new int[N][M];
        dp[0][0] = triangle[0][0];
        for(int i = 1; i < N; i++){
            for(int j = 0; j <= i; j++){
                if(j == 0){
                    dp[i][j] = dp[i-1][j] + triangle[i][j];
                }else if(j == i){
                    dp[i][j] = dp[i-1][j-1] + triangle[i][j];
                }else{
                    dp[i][j] += Math.max(dp[i-1][j-1], dp[i-1][j]) + triangle[i][j];
                }
            }
        }
        for(int i : dp[N-1]){
            answer = Math.max(answer, i);
        }
        return answer;
    }
}